storm-core/src/jvm/org/apache/storm/StormSubmitter.java

```diff
@@ -238,39 +239,24 @@ public static void submitTopologyAs(String name, Map stormConf, StormTopology to
                 }
 
                 // Dependency uploading only makes sense for distributed mode
-                List<String> jarsBlobKeys;
+                List<String> jarsBlobKeys = Collections.emptyList();
                 List<String> artifactsBlobKeys;
 
                 DependencyUploader uploader = new DependencyUploader();
                 try {
                     uploader.init();
                     jarsBlobKeys = uploadDependencyJarsToBlobStore(uploader);
                     artifactsBlobKeys = uploadDependencyArtifactsToBlobStore(uploader);
+                    setDependencyBlobsToTopology(topology, jarsBlobKeys, artifactsBlobKeys);
+                    submitTopologyInDistributeMode(name, topology, opts, progressListener, asUser, conf, serConf);
+                } catch (Throwable e) {
+                    // remove uploaded jars blobs, not artifacts since they're shared across the cluster
+                    uploader.deleteBlobs(jarsBlobKeys);
```

Comment:
InvalidTopologyException and AlreadyAliveException definitely indicates that the topology is not successfully deployed and respective jars can be removed from blob store.
 
@HeartSaVioR 
But this code assumes that any Exception thrown here indicates that the submitted topolgy is not deployed successfully.  I am not sure whether that assumption is always right.